oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -0,0 +1,268 @@
+     *         if the request failed.
+     * @throws TokenRequestException when the endpoint returns {code HTTP 400 (Bad Request)} status and the
+     *                               response payload contains the details of the error.
+     * @throws InvalidClientException when the endpoint returns {@code HTTP 401 (Unauthorized)} status, which
+     *                                typically indicates that client authentication failed (e.g.: unknown
+     *                                client, no client authentication included, or unsupported authentication
+     *                                method).
+     * @throws UnsupportedMediaTypeException if the media type of the response does not match the expected
+     *                                       (JSON).
+     */
+    private CompletableFuture<AccessTokenCapsule> refreshAccessTokenAsync(AccessTokenCapsule token) {
+        if (token.refreshToken() != null) {
+            // try refreshing token if refresh token was previously provided
+            return refreshRequest.make(token);
+        }
+        // try to re-obtain access token
+        return obtainAccessTokenAsync(token);
+    }
+
+    /**
+     * Refreshes access token using refresh token provided with the previous access token response, otherwise,
+     * if no refresh token available, re-obtains a new access token from the token end-point.
+     * If the refresh token request fails with {@link TokenRequestException}, tries to re-obtains a new access
+     * token from the token end-point.
+     * Optionally stores access token to registered {@link Consumer} for longer term storage.
+     * @return an {@link AccessTokenCapsule} that contains requested access token.
+     * @throws TokenRequestException when the endpoint returns {code HTTP 400 (Bad Request)} status and the
+     *                               response payload contains the details of the error.
+     * @throws InvalidClientException when the endpoint returns {@code HTTP 401 (Unauthorized)} status, which
+     *                                typically indicates that client authentication failed (e.g.: unknown
+     *                                client, no client authentication included, or unsupported authentication
+     *                                method).
+     * @throws UnsupportedMediaTypeException if the media type of the response does not match the expected
+     *                                       (JSON).
+     */
+    private AccessTokenCapsule refreshAccessToken(Instant instant) {
+        // after acquiring the lock, re-check if it's a valid token
+        final AccessTokenCapsule token = tokenRef.get();
+        if (token.isValid(instant)) {
+            // simply return a valid token
+            return token;
+        }
+        // otherwise, refresh it
+        AccessTokenCapsule refreshedToken;
+        try {
+            refreshedToken = refreshAccessTokenAsync(token).join();
```

Comment:
Question: Is there a special reason that we call `join()` here? Shouldn't we return `CompletableFuture<AccessTokenCapsule>`?

IIUC, `tokenRef` could be set by `refreshAccessTokenAsync(token).whenComplete(...)`

Reply:
Yes, there is a special reason for synchronizing the request execution here.
In order to deal with concurrency issues, all token `obtain` and `refresh` requests have to be executed serially by a dedicated `serialExecutor`, which was introduced based on the previous @trustin comment. So the obtaining or refreshing of the access token is indeed happens asynchronously by the `serialExecutor`, but anything that happens during the `refresh` or `obtain` have to be serialized.

Reply:
I've added a special test that tests the concurrency.

Reply:
I think we can compose asynchronous functions using `CompletableFuture.thenCompose` and process it sequentially.
```java
CompletableFuture<OAuth2AccessToken> oAuth2AccessToken =
    refreshAccessTokenAsync(token).thenCompose(this::obtainAccessTokenAsync);
```

Reply:
Sorry, I might have confused you with my explanation. This is roughly how it was implemented earlier, but this may cause concurrency issues obtaining or refreshing access token. The problem is the following: `getAccessToken()` method should be done asynchronously, but every such call has to be done sequentially in order to avoid concurrency issues accessing the OAuth 2 endpoint (the client may only obtain access token once and refresh the token when the time comes). As a result, I've introduced single-thread executor that can serialize asynchronous access token requests.
Imagine, there are two client requests sent to the protected resource simultaneously. Each will trigger `getAccessToken()` executed concurrently. Now, since the relevant chunks of `getAccessToken()` will be executed serially by the executor, the first request will obtain the access token, while the second request will re-use it and will not re-obtain the access token.
I've created `OAuth2ClientCredentialsGrantTest#testConcurrent()` to test such concurrent use case.

Reply:
I see, thanks for the kind explanation. Let me take a look `OAuth2ClientCredentialsGrantTest#testConcurrent()`. 😀

Reply:
If you consider that the two clients are sending `getAccessToken()` concurrently, how about using an async Caffeine cache?

```
AsyncCache<GrantedOAuth2AccessToken, CompletableFuture<GrantedOAuth2AccessToken>>
            accessTokenCache = Caffeine.newBuilder().maximumSize(1000 /* need sensible default */)
                                       .buildAsync();
accessTokenCache.asMap()
                .computeIfAbsent(accessToken, key -> {
                     // only one thread will be executed to obtain access token
                     // Note that we shold not put `RequestContextAwareFuture`
                     // need to peel `GrantedOAuth2AccessToken` and save it with CompletableFuture.completeValue(...)
                     return obtainAccessTokenAsync(key);
                });

```

I hope this workaround will solve the executor problem and concurrent issue.

Reply:
Looked at `ConcurrentMap#computeIfAbsent()`. Here is its code:
```
    default V computeIfAbsent(K key,
            Function<? super K, ? extends V> mappingFunction) {
        Objects.requireNonNull(mappingFunction);
        V oldValue, newValue;
        return ((oldValue = get(key)) == null
                && (newValue = mappingFunction.apply(key)) != null
                && (oldValue = putIfAbsent(key, newValue)) == null)
            ? newValue
            : oldValue;
    }
```
It does not actually guaranty that the `mappingFunction` (`obtainAccessTokenAsync`) will be executed only once, which we need in this case. Instead, it guarantees that the new value will be put only once.

Besides, even for a regular token read a hash lookup will take place, instead of simply reading a value. Therefore, we would add an additional cost to a frequent operation (of reading token) in sake of very rare operation (of obtaining new or refreshing token).

In any case, thanks for a suggestion, this was worth looking at.

Reply:
> It does not actually guaranty that the mappingFunction (obtainAccessTokenAsync) will be executed only once, which we need in this case. Instead, it guarantees that the new value will be put only once.

Yes, `ConcurrentMap` does not guarantee to execute only once but `Caffeine` cache uses `ConcurrentHashMap` internally.

`ConcurrentHashMap` override the implementation of `ConcurrentMap`.
```
    public V computeIfAbsent(K key, Function<? super K, ? extends V> mappingFunction) {
               ...
               Node<K,V> r = new ReservationNode<K,V>();
                synchronized (r) {
                    if (casTabAt(tab, i, null, r)) {
                        binCount = 1;
                        Node<K,V> node = null;
                        try {
                            if ((val = mappingFunction.apply(key)) != null)
                                node = new Node<K,V>(h, key, val);
                        } finally {
                            setTabAt(tab, i, node);
                        }
                    }
                }
    }
```
https://github.com/openjdk/jdk/blob/e427697e55c0d7b9089ba1f6ed24d4a313e9315b/src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java#L1703-L1713

After looking at Caffeine cache API, I realized that we don't need to call `asMap()`
It could be used like:
```
CompletableFuture<GrantedOAuth2AccessToken> future = cache.get(token, (key, executor) -> {
    // internally call ConcurrentHashMap.computeIfAbsent(...)
    return obtainAccessTokenAsync(key);
});

Reply:
Unfortunately, the `AsyncCache` does not use `java.util.concurrent.ConcurrentHashMap#computeIfAbsent` directly. Instead, it has its own `com.github.benmanes.caffeine.cache.LocalCache#computeIfAbsent` API and implementations like `com.github.benmanes.caffeine.cache.BoundedLocalCache#computeIfAbsent`. I did not perform and deep code analysis or conducted any hands-on testing, but as far as I can see, it only guarantees a consistency of the operation on `LocalCache<K, CompletableFuture<V>>`. Although the future instances will be obtained serially, it does not prevent two `mappingFunction`s to execute simultaneously, at least I don't see such mechanism. If you have better insight, please explain me how it's achieved.

Please also comment on the why with this Cache approach "we would add an additional cost to a frequent operation (of reading token) in sake of very rare operation (of obtaining new or refreshing token)". Cache lookup will have to be executed for each outgoing request vs simple `AtomicReference#get()`. For thousands of requests it could be significant cost.

Reply:
> Unfortunately, the AsyncCache does not use java.util.concurrent.ConcurrentHashMap#computeIfAbsent directly. Instead, it has its own com.github.benmanes.caffeine.cache.LocalCache#computeIfAbsent API and implementations like com.github.benmanes.caffeine.cache.BoundedLocalCache#computeIfAbsent.

- `BoundedLocalCache` keeps its data on `ConcurrentHashMap`. You can easily check it by breaking point on the `mappingFunction`.
https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java#L213

> Cache approach "we would add an additional cost to a frequent operation (of reading token) in sake of very rare operation (of obtaining new or refreshing token)". 

The reason why I suggested to use cache is:
- We can remove the serial executor. It is hard to tune. 
  - If `AbstractOAuth2AuthorizationGrant` has a static singleton executor, then it could have trouble with the scalability problem. Because the submitted tasks are blocking IO.
  - If `AbstractOAuth2AuthorizationGrant` has its own executor, users might not handle many resource owners
- Armeria is a non-blocking microservice framework. If it is possible, we don't have to block IO.
- The verification logic on `tokenRef` is pretty difficult for me. I think we can hide the logic.

> Cache lookup will have to be executed for each outgoing request vs simple AtomicReference#get(). For thousands of requests it could be significant cost.

Yes, that's true. Hash lookup operation is expensive that `AtomicReference`. However, I think the context switching cost is more expensive than them.

Reply:
Just to help clarify (regardless of your final decisions), Caffeine does perform the `Map.compute` calls atomically. In the case of `computeIfAbsent` it is actually a fast-path `ConcurrentHashMap#get` followed by a `ConcurrentHashMap#compute` if absent. This allows us to optimistically return a valid value cheaply, whereas `ConcurrentHashMap#computeIfAbsent` pessimistically locks which limits read concurrency. We fallback to a compute in order to handle cases like expired entries still in the map and similar caching details.

The `AsyncCache` will perform the map operations atomically, but only establishes the mapping for the future. The future is therefore independent, so atomically established and asynchronously completed. If the future fails (exceptionally or returns a null result), then a `whenComplete` handler will remove it from the cache.

The simplest way to think about Caffeine is as `ConcurrentHashMap` with extra features, which tries as much as possible to not deviate. There are very small differences, primarily that our `#clear` cannot block on unestablished mappings hidden by the iterator, whereas CHM does by traversing its internal table. But from atomicity standpoint we try hard to be correct while also performant. A `Cache` ~= `CHM`, `AsyncCache` ~= `CHM<K, CF<V>>`, and the `asMap` views let you use the collections api rather than a cache-oriented one.

The JavaDoc can always be improved, but does try to convey this atomicity. If you have any suggestions to clarify it, please let me know.

As an aside, I often use [Failsafe](https://github.com/jhalterman/failsafe) for failure handling, retries, etc. You could avoid the `join` by using an async fallback so that the returned future encapsulates this error handling logic. That avoids surprises like starving FJP when its workers are all blocked waiting on tasks queued for execution. I am assuming that you'd prefer returning an async result rather than a synchronous API.

```java
var fallback = Fallback.ofStage(this::obtainAccessToken)
    .handleIf(failure -> Exceptions.peel(e) instanceof TokenRequestException);
return Failsafe.with(fallback).getAsync(() -> refreshAccessTokenAsync(token));
```

Reply:
Thanks a lot for chiming in, @ben-manes!

+1 for going full-async rather than using a dedicated executor.

Reply:
Thank you @ben-manes for clarification!

Let me comment on some points...

>> The AsyncCache will perform the map operations atomically, but only establishes the mapping for the future. The future is therefore independent, so atomically established and asynchronously completed.

This is actually the key part of the clarification in our context. And this is how I interpreted the `AsyncCache` code.
In our case the mapping function have to be executed serially, that's the only way to guarantee that Obtain or Refresh operation **executed only once**.

>> Armeria is a non-blocking microservice framework. If it is possible, we don't have to block IO.

I think I can understand that. That's why Obtain/Refresh token operation executed asynchronously by a different (executor) thread.

>> The verification logic on tokenRef is pretty difficult for me. I think we can hide the logic.

Apparently not behind the `AsyncCache`, as explained above.
Let me summarize some concrete requirements so you can advise better:

- `getAccessToken()` shall be asynchronous;
- `getAccessToken()` shall be extremely cheap when the access token already issued;
- When the token not yet available or expired, `Obtain` or `Refresh` access token shall take place only once - no duplicate requests should be issued towards the OAuth 2 endpoint; 

I'm not saying the solution with the `Executor` is perfect, but at the moment I only see `Queue` or `Executor` approach can reasonably facilitate the three requirements above. 
If you just don't like seeing such complexity inside the `Grant`, this can easily be achieved by extracting it into some more generic utility class.

>> Yes, that's true. Hash lookup operation is expensive that AtomicReference. However, I think the context switching cost is more expensive than them.

Not sure I understand where the context switching takes place (as oppose to the hash lookup)? In regular case `AtomicReference#get()` is all we are going to have. Once in a while, on a rare occasion (either when there is no access token issued yet or when the access token to be renewed) the processing will take its turn through the serial executor. This minimizes the negative impact comparing to frequent hash lookup taking place for each request.
For the Obtain/Refresh operations context switching is unavoidable and must be facilitated by a different thread to avoid blocking the event loop. These operations may take significant time to complete (OAuth 2 endpoint will perform authentication, authorization, security artifact generation and provisioning, etc). Some operations may take much longer and involve continuous polling of the OAuth 2 token endpoint (e.g. [Device Authorization Flow](https://tools.ietf.org/html/rfc8628#section-3.5), especially `authorization_pending` and `slow_down` type of responses - this one I was planning to add next as the code already developed).